Initial graph implementation

JoOkuma opened the pull request:
Hey,

This is just a naive pure python implementation to get things started.
I'm planning to benchmark with other data structures, more operations (e.g. iterating over the graph), and other packages (e.g. numba).

Should we support edge features on this first implementation?

How should we organize the files?

JoOkuma commented:
I implemented the trackmate-like graph, it's very fast using numba.
I had to duplicate the number of edges for the undirected graph, I'm not sure if there's a better way to do this.

jni commented:
Very cool stuff @JoOkuma!

Mild comments:

- I would suggest having a single Graph class, and then directed, undirected, and multigraphs can either have an internal flag, or inherit from the parent class.
- I think you do need to duplicate edges for undirected graphs. If you're only ever checking for edge presence, then you can always check (u, v) in E or (v, u) in E. But for things like traversal, this is not an option, and you'd be stuck checking whether *any* other node has an incoming edge to your current node, which is slow. *However*, when it comes to e.g. edge features, you can always give (u, v) and (v, u) the same ID, and they resolve to the same feature row.
- I'm happy to help with package structure. I'll do that in a separate PR and then you can rebase?
- I'd prefer to just use numba as numba rather than nb, as it is less common than the other packages. Also you don't need to call `njit`: `@numba.njit` (no parens) will work.
- Should we start with 2x the number of node/edge slots rather than 1x? Or 1.3x? Something that won't immediately need a reallocation if we click "add". :joy: 

Exciting to see this taking shape!

jni commented:
@JoOkuma see #2. Let me know if it looks ok to you and I'll merge.

JoOkuma commented on src/benchmark_undirectedgraph.py:
```diff
@@ -0,0 +1,78 @@
+    return nodes_df, edges
+
+
+def from_networkx(nodes_df, coordinates_columns, edges) -> nx.Graph:
```
indeed, I'll fix that on the next commit.

JoOkuma commented:
> I would suggest having a single Graph class, and then directed, undirected, and multigraphs can either have an internal flag, or inherit from the parent class.

Ok, I want to benchmark adding, and transversing the graphs, and then I'll clean up to a single graph class.

>   However, when it comes to e.g. edge features, you can always give (u, v) and (v, u) the same ID, and they resolve to the same feature row.

Good idea, sibling edges could always be next to each other (e.g, `i, i+1`), so we can index the features by `i // 2`, delete the pair without having to search for the sibling, and ignore duplicates fast when sending the bufffer to vispy `[::2]`.

>  I'd prefer to just use numba as numba rather than nb, as it is less common than the other packages. Also you don't need to call njit: @numba.njit (no parens) will work.

👍

>   Should we start with 2x the number of node/edge slots rather than 1x? Or 1.3x? Something that won't immediately need a reallocation if we click "add". 😂

Great! It could be an optional parameter.

JoOkuma commented:
@jni or anyone else, any thoughts on nodes' indexing?

I think the user should manage the node indexing and we only return the buffer index and keep it consistent.
For example, the node `add_nodes(<nodes' coordinates/features>)` would return the indices of the added nodes. This is advantageous to us because we don't have to worry about arbitrary indexing and keep things efficient by using our buffers' index.
The downside is if the user decides to use their own indexing they'll have to remap their indices to do any operation, **including adding edges**.

If we decide to let the user provide their own indexing then we need to store a dictionary from their indexing to our buffer's index and we might not get that much of a speedup. I'm not sure if there's another way to implement this.

jni commented:
@JoOkuma sorry about the slow response! But at any rate, I don't like the idea of not providing indexing. It's weird to provide nodes with only coordinates. Yes you get indices, but as a user I want to provide a graph, which is usually a list of edges, which are id pairs. giving the graph as coordinate pairs is obviously fraught (since coordinates are floating point numbers), while adding the graph in two stages is awkward: here's my point coordinates, now give me my ids, ok now I need to figure out how the coordinates I gave you match those ids and translate my edges to give you the edges... It's all very messy.

With regards to speed, I think it's ok: we have a mapping from external node IDs to sequential IDs as a dictionary; numba supports int-int dictionaries and we only use this rarely: internally, all the operations are on sequential IDs, and when we need to ask for original IDs, the *reverse* mapping can be stored in an array, which is very fast.

I think operations like adding and removing edges can be "quite fast", but don't need to be blazing fast. The critical speed issues are traversing paths and traversing the whole graph, which we can do in sequential ID land.

Or, at least, that's my rationale. I think the alternative API is too horrible to contemplate though, so we should give this a go. :joy: *If* we measure a hideous bottleneck in the remapping, we can reconsider.

JoOkuma commented:
It makes a lot of sense and we restrict the API to integer indexing only.

I implemented a few additional features and I'll get back to this next week, I need to finish a few other things this week.

JoOkuma commented:
@jni, it's taking shape.

My numba implementation seems very C like, euheuehu. I was hoping for a better way to deal with the constants while still inlining them but I couldn't figure it out.

JoOkuma commented:
I think the basics are mostly done.

Some points that are still missing are:

- [ ] A better approach to handle the nodes' features. I need to check how the feature tables are used in napari.
- [ ] Slicing the graph for visualizing it in napari --- important for tracks layer.
- [ ] A function serializes the graph data to a vispy buffer.

jni reviewed (commented):
@JoOkuma this is super awesome! I have pushed some formatting changes to match Black and PEP257 and numpydoc, and then made a few comments. I intend to keep reviewing but I have to go right now and didn't want to leave this review pending. 👍 Thank you!

JoOkuma commented on src/napari_graph/_base_graph.py:
```diff
@@ -0,0 +1,652 @@
+class BaseGraph:
+    """Abstract base graph class.
+
+    Parameters
+    ----------
+    n_nodes : int
+        Number of nodes to allocate to graph.
+    ndim : int
+        Number of spatial dimensions of graph.
+    n_edges : int
+        Number of edges of the graph.
```
I agree with you, since we will only support spatial graphs we need to provide the coordinates, so it could be `Graph(my_edges_sparse_matrix, node_indexed_coordinates_dataframe)`. Using a pandas data frame with a sparse matrix adds some flexibility with the indexing, rather than having a coordinate numpy array.

jwindhager replied:
Hi @JoOkuma and @jni, was wondering where we stand on this topic? Currently, napari-graph stores `coords`. Would you rather move this to napari's Graph layer, @jni?

JoOkuma commented on src/napari_graph/directed_graph.py:
```diff
@@ -0,0 +1,430 @@
+    edges_buffer: np.ndarray,
+    node2src_edges: np.ndarray,
+    node2tgt_edges: np.ndarray,
+    is_target: int,
```
I use to shift the index of the source/target linked list `next_idx = edges_buffer[buffer_idx + _LL_DI_EDGE_POS + is_target]` inside the numba function.

jni replied:
Ok, maybe make a note in the docstring? 😊 Even for private functions, it's useful to document every parameter.

JoOkuma commented on src/napari_graph/directed_graph.py:
```diff
@@ -0,0 +1,430 @@
+)
+from napari_graph.undirected_graph import _UN_EDGE_SIZE
+
+# directed edge constants
+_DI_EDGE_SIZE = _UN_EDGE_SIZE + 1
+_LL_DI_EDGE_POS = 2
```
Added explanation for these variables and for the ones on the undirected graph.

jni commented on src/napari_graph/directed_graph.py:
```diff
@@ -0,0 +1,418 @@
+
+    NOTE: see `_add_undirected_edge` docs for comment about cache misses.
+
+     Returns
```
```suggestion
    Returns
```

It'd be good to add parameter descriptions too — as noted elsewhere, this can be very useful for someone coming in to the code later. (Even your later self 😉)

jni commented on src/napari_graph/directed_graph.py:
```diff
@@ -0,0 +1,418 @@
+    idx = node2tgt_edges[tgt_node]  # different indexing from source edge
+    prev_buffer_idx = _EDGE_EMPTY_PTR
+
+    while idx != _EDGE_EMPTY_PTR:
```
As noted elsewhere, can this be converted to a for loop to ensure that it'll terminate? One example could be `for i in range(edges_buffer.shape[0]):`, since it's guaranteed that you will have looked at every position by then *or* you'll be in an infinite loop. ;)

jni commented on src/napari_graph/directed_graph.py:
```diff
@@ -0,0 +1,418 @@
+    else:
+        idx = node2src_edges[node]
+
+    while idx != _EDGE_EMPTY_PTR:
```
As above — avoid bare while loops.

jni commented on src/napari_graph/undirected_graph.py:
```diff
@@ -0,0 +1,306 @@
+    # numpy vectorization
+    idx = node2edges[node]
+
+    while idx != _EDGE_EMPTY_PTR:
```
As above — avoid bare while loops, make sure it's a for-loop, and maybe raise an error if you reach the end

jni reviewed (commented):
@JoOkuma sorry that I'm so slow to react here, but I've left a few more comments around the implementation. Summary:

- please avoid while loops — even if you *think* you've guaranteed that it will terminate, and *even* if you make no errors, nothing prevents a user or a plugin from corrupting the table, and then you might get an infinite loop. Better to have a for-loop and raise a value error if you've taken more steps than should be possible.
- can we make the public API as close as possible to networkx? Or do you think there are reasons to avoid that?
- As noted in an inline comment, I think *napari* will concern itself only with spatial graphs, but I think this library can probably remain more general? That is, this library can and should support graphs without spatial information. When adding a graph to napari, though, we require nodes to have coordinates of the appropriate dimensionality.

JoOkuma commented:
@jni 

> loops please avoid while 

I will try to fix that.

> can we make the public API as close as possible to networkx? Or do you think there are reasons to avoid that?

Yes, but it will be less consistent with the naming convention used in the trackmate's white paper, the backend should remain the same, though.

> I think napari will concern itself only with spatial graphs, but I think this library can probably remain more general?

Yes, we can, so the constructor input will be only a sparse matrix.
I discussed a bit with the folks from the architecture working group and I'll remove the `features` from the napari-graph, the layer can manage that, I think the same could be done with coordinates, I'm going to think if there's any shortcoming.

JoOkuma commented:
@jni, the downside of removing the coordinates from `napari-graph` is that the user will have to manage it on its own, which could be more inefficient.
This won't affect the graph layer implementation because I'm only accessing the `_coords` buffer as it's more efficient and it shares the functionality from the points layer, [reference](https://github.com/JoOkuma/napari/blob/dd1a5209d48b53d5f6d9814b68271956d55dc804/napari/layers/graph/graph.py#L145-L148).
I'm going to proceed with removing this, but we should keep this in mind.

JoOkuma commented:
Still, getting the hang of GitHub, should I "add a review" to your comments or just resolve them?

Tomorrow I'll try to finish the API change to scipy sparse matrices and remove the coordinates.

JoOkuma commented:
The tests were not running on the CI, kkk.

I'll go back to the layer now.

The recent changes were:
- added coverage
- remove `feats`
- update constructor
- added some alias on the directed graph to be more similar to networkx
- update constructor, I decided to use a list of edges rather than a sparse matrix, the sparse matrix would lead to some complications for non-consecutive node indexing or an empty graph.

jwindhager commented on src/napari_graph/directed_graph.py:
```diff
@@ -0,0 +1,498 @@
+            node2edges=self._node2edges,
+            iterate_edges_func=_iterate_directed_source_edges,
+            mode=mode,
+        )
+
+    def in_edges(
+        self, nodes: Optional[ArrayLike] = None, mode: str = 'indices'
+    ) -> Union[List[np.ndarray], np.ndarray]:
+        """`target_edges` alias"""
+        return self.target_edges(nodes, mode)
+
+    def target_edges(
+        self, nodes: Optional[ArrayLike] = None, mode: str = 'indices'
+    ) -> Union[List[np.ndarray], np.ndarray]:
+        """Return the target edges (incoming) of the given nodes.
+
+        If no nodes are provided, all target edges are returned.
+
+        Parameters
+        ----------
+        nodes : Optional[ArrayLike], optional
+            Node indices, by default None
+        mode : str
+            Type of data queried from the edges. For example, `indices` or
+            `coords`.
+
+        Returns
+        -------
+        List[np.ndarray]
+            List of (N_i) x 2 x D arrays, where N_i is the number of edges at
+            the ith node.  D is the dimensionality of `coords` when
+            mode == `coords` and it's ignored when mode == `indices`. N_i
+            dimension is ignored when N_i is 1.
+        """
+        return self._iterate_edges_generic(
+            nodes,
+            node2edges=self._node2tgt_edges,
+            iterate_edges_func=_iterate_directed_target_edges,
+            mode=mode,
+        )
```
For maintenance reasons, personally, I'd prefer to not introduce aliases. Don't have strong opinions regarding source/target vs in/out, though.

jwindhager commented on src/napari_graph/_base_graph.py:
```diff
@@ -0,0 +1,749 @@
+
+        self._add_edges(edges)
+
+    @abstractmethod
+    def _remove_edges(self, edges: np.ndarray) -> None:
+        raise NotImplementedError
+
+    def remove_edges(self, edges: ArrayLike) -> None:
+        """Remove edges from the graph.
+
+        Parameters
+        ----------
+        edges : ArrayLike
+            A list of 2-dimensional tuples or an Nx2 array with a pair of
+            node indices.
+        """
+        edges = self._validate_edges(edges)
+        edges = self._map_world2buffer(edges)
+        self._remove_edges(edges)
+
+        # FIXME: this can lead to inconsistency in the count if removing edges
+        # raises an error
+        self._n_edges -= len(edges)
```
Partially addressed by https://github.com/JoOkuma/napari-graph/pull/8

jwindhager replied:
This has been further discussed in https://github.com/JoOkuma/napari-graph/pull/8#issuecomment-1358767828, resolving

jwindhager commented on src/napari_graph/_base_graph.py:
```diff
@@ -0,0 +1,749 @@
+            self.init_data_from_dataframe(coords)
+
+        if len(edges) > 0:
+            self.add_edges(edges)
```
See https://github.com/JoOkuma/napari-graph/pull/9

jwindhager commented on src/napari_graph/_base_graph.py:
```diff
@@ -0,0 +1,749 @@
+        if len(edges) > 0:
+            self.add_edges(edges)
+
+    def _init_buffers(self, n_nodes: int, n_edges: int) -> None:
+        # node-wise buffers
+        self._empty_nodes: List[int] = list(reversed(range(n_nodes)))
+        self._node2edges = np.full(
+            n_nodes, fill_value=_EDGE_EMPTY_PTR, dtype=int
+        )
+        self._world2buffer = typed.Dict.empty(types.int64, types.int64)
+        self._buffer2world = np.full(
+            n_nodes, fill_value=self._NODE_EMPTY_PTR, dtype=int
+        )
+        # edge-wise buffers
+        self._empty_edge_idx = 0 if n_edges > 0 else _EDGE_EMPTY_PTR
+        self._n_edges = 0
+        self._edges_buffer = np.full(
+            n_edges * self._EDGE_DUPLICATION * self._EDGE_SIZE,
+            fill_value=_EDGE_EMPTY_PTR,
+            dtype=int,
+        )
+        self._edges_buffer[
+            self._LL_EDGE_POS : -self._EDGE_SIZE : self._EDGE_SIZE
+        ] = np.arange(1, self._EDGE_DUPLICATION * n_edges)
```
> Maybe split up into `_init_node_buffers` and `_init_edge_buffers`? Happy to submit a PR if desired.

See https://github.com/JoOkuma/napari-graph/pull/9

> Also, maybe rename as follows (mainly for consistency)?

For this, I'll wait for feedback from you & for the PR above

jwindhager replied:
Hi @JoOkuma 

Re the renaming: the reason I suggested this was to make the naming of variables a bit more consistent, but this isn't restricted to just this part of the code. There are a number of places in the current implementation where it isn't instantly clear whether e.g. an *edge* is supposed to be a 2-tuple, an *edge number*, or an *edge buffer index*. Same goes for *node*, where it e.g. isn't clear from the current naming scheme that this in many cases is a *node index* and not a *node label*.

Trying to come up with a more generalized terminology, I propose the following naming scheme (hopefully capturing most of the information floating around):
- *node index*, i.e. an integer in the range `[0, n_nodes)` (also directly serves as *node buffer index*)
- *node label*, i.e. a user-defined unique node identifier (mapped to the corresponding *node index* via `world2buffer`/`buffer2world` arrays in the current implementation, if I understand correctly)
- *node coordinate*, i.e. a `d`-tuple of floats (mapped via *node index*)
- *edge*, i.e. a unique 2-tuple of *node indices* (not *node labels* (!) if I understand correctly)
- *edge number*, i.e. an integer in the range `[0, n_edges)`
- *edge buffer index*, i.e. an integer in the range `[0, len(edges_buffer))`

... and to refactor the code accordingly to make the names of attributes, arguments, etc a bit more clear/expressive/consistent. What do you think?

Happy to give this a try, but it may turn out to be quite a "busy" PR I guess :-)

jwindhager commented:
Just noticed: I would rename `BaseGraph` to `Graph` and make it part of the public interface. This can for example be useful for typing during plugin development.

JoOkuma commented:
> Just noticed: I would rename `BaseGraph` to `Graph` and make it part of the public interface. This can for example be useful for typing during plugin development.

I agree it's useful.
I'm afraid some people would try using `Graph` and get the abstract class error without trying the other classes.

Do you think renaming `_BaseGraph` to `BaseGraph` and making it public is a good compromise or it's about the same as having it named `Graph`?

JoOkuma commented:
@jwindhager what do you think is missing for us to merge this PR?

I'm going to write a few test cases for a graph without coordinates.

jwindhager commented:
Hi @JoOkuma 

From my point of view:
- [ ] As you write: ensure that the graph works without coordinates
- [ ] Make `BaseGraph` part of the public API (I agree with your suggestion!)
- [ ] Improve error handling during adding/removal, see https://github.com/napari/napari-graph/pull/1#discussion_r1049728975 and https://github.com/JoOkuma/napari-graph/pull/8#issuecomment-1358767828
- [ ] Address the remaining comments of @jni (if any?) and in my review comments above
- [ ] Some more tests and documentation (incl. more verbose docstrings)

Happy to help, will continue to file PRs for you to review! :-)

JoOkuma commented:
Could we leave the documentation to another PR?

jwindhager commented on src/napari_graph/_base_graph.py:
```diff
@@ -0,0 +1,749 @@
+        """
+        node_indices = self._validate_nodes(node_indices)
+        node_indices = self._map_world2buffer(node_indices)
+        return self._coords[node_indices]
```
https://github.com/JoOkuma/napari-graph/pull/12

jwindhager commented on src/napari_graph/_base_graph.py:
```diff
@@ -0,0 +1,749 @@
+        self._world2buffer[index] = buffer_index
+        self._buffer2world[buffer_index] = index
+
+    def _alloc_size(self, size: int) -> int:
```
https://github.com/JoOkuma/napari-graph/pull/12

kephale commented:
@JoOkuma @jwindhager @jni, I'm planning on digging into `napari-graph` while at the Janelia Cell Tracking hackathon (starting tomorrow for me). 

If you have any thoughts, requests, etc., then please let me know.

JoOkuma commented:
Thanks, @kephale .

@jwindhager What do you think about merging this PR so we can improve it with other PRs?
This PR is huge already, and @jwindhager was (almost) done reviewing it.

I think not having this merged into main makes it more difficult for others to contribute.

jni commented:
Agreed @JoOkuma! 😊 🎉 🚀 here we go!

JoOkuma commented:
Thanks, @jni

jwindhager commented:
Hey, apologies for the late reply! Yes, merging was definitely the right way to go forward :) Thanks @jni!
